File: Best Time to Buy and Sell Stock IV.py
class Solution:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        #Approach: One-pass Simulation  //  refer LeetCode #123 for explanation
        #Time Complexity: O(n * k)
        #Space Complexity: O(k)
        #where, n is the length of prices
        
        if not prices or k == 0:
            return 0
        
        buyK = [inf for _ in range(k)]
        profitK = [0 for _ in range(k)]
        
        for price in prices:
            buyK[0] = min(buyK[0], price)
            profitK[0] = max(profitK[0], price - buyK[0])
            
            for i in range(1, k):
                buyK[i] = min(buyK[i], price - profitK[i - 1])
                profitK[i] = max(profitK[i], price - buyK[i])
                
        return profitK[-1]

File: best-time-to-buy-and-sell-stock-iv.cpp
//Time - O(nxk)
//Space - O(k)
class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        vector<int> buy (k+1, INT_MAX);
        vector<int> sell (k+1, 0);
        
        for(auto p:prices){
            for(int i = 1;i<=k;i++){
                buy[i] = min(buy[i], p - sell[i-1]);
                sell[i] = max(sell[i], p - buy[i]);
            }
        }
        
        return sell[k];
    }
};